client/tx/tx.go

```diff
@@ -310,6 +311,65 @@ func PrepareFactory(clientCtx client.Context, txf Factory) (Factory, error) {
 	return txf, nil
 }
 
```

Comment:
The two new functions below are very verbose. For one, we can shorten the signature by omitting duplicate types, e.g. `uint64`. Also, it reads pretty weird to have both a TxGenerator and a TxBuidler (which is being removed AFAIK?), can we not just use one? Does one not already have everything we need (e.g. chain ID, nonce, account num, etc...)?

Reply:
> it reads pretty weird to have both a TxGenerator and a TxBuilder (which is being removed AFAIK?)

A TxGenerator can create TxBuilders (see adr 20), so I don't think that's removed:

https://github.com/cosmos/cosmos-sdk/blob/6f928e1c4dffe0bec51736fca7fd5e8a5e7570d7/client/tx_generator.go#L12-L15

What's removed is the old TxBuilder in x/auth/types:

https://github.com/cosmos/cosmos-sdk/blob/6f928e1c4dffe0bec51736fca7fd5e8a5e7570d7/x/auth/types/txbuilder.go#L18-L20

--- 

> Does one not already have everything we need (e.g. chain ID, nonce, account num, etc...)?

The struct that has everything we need is `Factory` (which internally holds a `client.TxGenerator`). I can refactor the tests to use Factory, should help in readability.

Initially, I was hoping ideally the tests in x/auth/ante not to use `client` at all, because they are just ante handler unit tests after all. But that's maybe too big a refactor, I'll try with Factory and see how it goes.

Reply:
After coming away from this work for a bit, it really seems messy TBH. Builder, Factory, generator...all seems like a codesmell.

Reply:
Ahh right, so I meant the _original_ `TxBuilder` was being removed -- I completely forgot we had an interface. Taking a look again, is there any reason we can't just combine the `TxBuilder` and `TxGenerator` interfaces into one single interface -- `TxGenerator`? Thoughts @aaronc?

Reply:
TxGenerator and TxBuilder can’t be the same thing. TxBuilder wraps a transaction. TxGenerator can create a transaction as well as do other things like encoding and managing sign modes. So I don’t think we can remove a type there but maybe a renaming would help. TxGenerator should probably be called TxContext or something. I actually think the API as it is mostly makes sense. 

Factory is a bit confusing and imho overlaps with client.Context so if l were to do anything l might combine those two types. Factory mainly reads parameters from viper afaik so it seems like maybe all of that stuff should be on client.Context and read in via flags.

Reply:
I would certainly not combine Factory with Context. I think the APIs are confusing IMHO. Generator and Builder sound like the same thing. We should at least rename one of those.

Reply:
Between tx construction and broadcast there are only ever 5 steps:
1. input tx-specific params
2. generate the signing bytes
3. sign those bytes
4. construct the final tx (with signature) & encode it
5. broadcast

> TxBuilder wraps a transaction

Right now, TxBuilder only holds: Msg, Fee, GasLimit, Memo. Would it make sense, as a 1st step, **to remove Factory** and put all its relevants fields (chainID, accountNum, sequence, aka `SignerData`) inside TxBuilder? 

https://github.com/cosmos/cosmos-sdk/blob/e7554bb3b0052b734528a85adbead837ed9d6dca/client/tx/factory.go#L18-L29

This way, TxBuilder handles step 1 above.

---

@aaronc if i understand you correctly, TxGenerator takes care of steps 2 (via `SignModeHandler().GetSignBytes`) and 4 (via `TxEncoder`). Since both steps are generic and can be customized by end-user, I think I understand why it needs to be a separate interface. But yeah, TxGenerator is a confusing name... maybe TxSignEncode?

Both functions should receive `TxBuilder` as arg.

---

Regarding step 3, right now we have:
https://github.com/cosmos/cosmos-sdk/blob/a65bcdd7e551ffc8414f58d85f00d74295a42864/client/tx/tx.go#L321

I'd just replace the Factory arg by a keyring.Keyring or crypto.PrivKey

Reply:
I would rename `TxGenerator` either `TxContext` of `TxConfig`. I think `TxConfig` makes sense because it is the tx configuration setup in the chain - either amino or protobuf as well as the supported sign mode handlers.

Reply:
Let's keep in mind that `TxBuilder` wraps an actual transaction. So I don't really think there's much need to change it unless what is happening in the underlying transactions changes.

`Factory` is primarily related to reading parameters from the CLI. It's a very different type of concern. `TxGenerator` and `TxBuilder` have nothing to do with the CLI and they shouldn't.

In my mind the biggest overlap is between `Factory` and `client.Context`. They have a similar concern - reading parameters from the CLI and actually share several fields already (`Keyring`, `TxGenerator`, `AccountRetriever`, `ChainID`). `Factory` however extends what's in `client.Context` to also read the tx CLI params like gas, fees, memo, etc. I'm fine with keeping both types but I wonder what your rationale for separating them is @alexanderbez ? Maybe a renaming would be helpful there too.

Reply:
> I would rename TxGenerator to either TxContext of TxConfig. 

I'm in favor of this.

Reply:
> I just really want to make sure `client.Context` doesn't balloon into something that handles many orthogonal things.

I totally agree. I actually think `client.Context` is already too big (> 20 fields) and IMHO I would prefer more smaller, well-named types for this CLI stuff rather than one big catch-all type. In `client.Context` there are two type of things:
* app config (`JSONMarshaler`, `InterfaceRegistry`, `TxGenerator` and `AccountRetriever`). These always need to be populated by the app and are never read from flags
* things that get initialized by flags
* `Input` and `Output` which relate to cobra/testing and maybe are already covered by `*cobra.Command`??

I think the things that get initialized by flags should be put in flag sets. There are the flags that apply to all commands (like `NodeURI`) and the things that apply to only transactions. So basically query flags and tx flags or something like that. Anyway, maybe we're getting into a larger discussion here... but CLI stuff could probably be structured more cleanly...

Reply:
Fine by me @amaurymartiny.

Reply:
Will also use Factory here, this should increase readability and remove the need of these 2 new functions here.